### PR TITLE
Fix crashes with WorkspaceGroup input to algorithms

### LINKED ISSUE #15433: Add unit test for using AsymmetryCalc with WorkspaceGroups
Add a regression test for the bug fix in #15404. Test that the algorithm can be run on a WorkspaceGroup, that it runs once on each workspace in the group and that the outputs are then grouped.

### LINKED ISSUE #15575: CalMuonDetectorPhases: crashes with workspace groups
The exact same issue as #15402 looks like it also happens with this algorithm. The algorithm does not cope properly with input WorkspaceGroups.

To replicate:
- load a file such as `EMU00006475.nxs`, which is loaded into a WorkspaceGroup (run with multiple periods)
- Highlight it in the workspace selector and run `CalMuonDetectorPhases` from the GUI algorithm window
- Mantid crashes

As part of this issue, check other algorithms for the same behaviour as this could be in a few places!

### PR DESCRIPTION
**Pull request by tom-perkins**
Several algorithms turned out to have the same bug in `validateInputs()` that `AsymmetryCalc` was found to have in #15402. The input workspace was assumed to be a `MatrixWorkspace` or similar, and so the algorithm crashed when given a `WorkspaceGroup` as input. 

This PR fixes all the cases I could find, and adds tests that expose the bug and verify the fix:
- `AsymmetryCalc` (fixed in #15404 but this adds a test)
- `CalMuonDetectorPhases`
- `ConvertToDistribution`
- `ChangeTimeZero`
- `FFT`
- `MaxEnt`

**To test:**

_Code review:_
1. Try to find cases of this crash that I've missed - an algorithm with a `validateInputs()` method that uses the input workspace in its validation, and fails to check for null.
2. In some of the cases, would it be better for the algorithm to explicitly reject WorkspaceGroups rather than running once on each workspace?

_Functional tests:_
The new unit tests are intended to test the fixes here.
 To test manually, create a WorkspaceGroup in Mantid, highlight it in the MantidPlot workspace selector and then try to run the algorithm via the GUI. Mantid should not crash. (See description of #15402 for an example)

Fixes #15575 
Fixes #15433 

[Release notes](https://github.com/mantidproject/mantid/blob/15575_workspace_group_crashes/docs/source/release/v3.7.0/framework.rst) have been updated.

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the coding standards? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [x] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/test/CalMuonDetectorPhasesTest.h [hunk 1]
```diff
@@ -98,6 +112,34 @@ class CalMuonDetectorPhasesTest : public CxxTest::TestSuite {
...
+    AnalysisDataService::Instance().add("workspace1", ws1);
+    AnalysisDataService::Instance().add("workspace2", ws2);
+    auto group = boost::make_shared<WorkspaceGroup>();
+    AnalysisDataService::Instance().add("group", group);
+    group->add("workspace1");
+    group->add("workspace2");
+    TestCalMuonDetectorPhases calc;
+    calc.initialize();
+    calc.setChild(true);
+    TS_ASSERT_THROWS_NOTHING(calc.setPropertyValue("InputWorkspace", "group"));
```
Comment (FedeMPouzols) [inline Framework/Algorithms/test/CalMuonDetectorPhasesTest.h]: I think that here you could use `setProperty()` and pass the workspace group object itself, so you wouldn't need to add the workspaces in the "ADS".

**Comment by FedeMPouzols** [comment]
Code changes looks good, and the release notes list all the algorithms that have been improved.
Fixes a bunch of algorithms that have issues with workspace groups. Fixes the `validateInputs()` methods and adds new test cases for workspace groups.

The other `validateInputs()` that I saw in Framework look safe. They either do not check workspace properties or do properly check against nullptr the workspace pointer after the `getProperty()`.
I think it would be really difficult to be 100% sure that all algorithms that could have this issue are now behaving correctly and/or can cope with groups. My guess is that support for workspace groups is not provided consistently across all areas. Some techniques will naturally tend to use groups very often, and probably the relevant algorithms can validate and process groups adequately. Related to this, I'd also suspect that some algorithms may have issues when they are given wrong workspace types, not only groups (event / matrix / etc.) and they do a `getProperty()` assuming a particular type of workspace. If the `getProperty()`s were done via casting coverity and other tools could help, but with the current mechanism I don't see an easy automated way of catching these issues.

But this definitely fixes the issue for several algorithms, and I couldn't find any obvious additional candidate for similar fixes.

@tom-perkins : we have this wiki page http://www.mantidproject.org/Writing_an_Algorithm where at the end they explain validators and `validateInputs()`. Maybe we could add a note there to remind ourselves to always consider workspace groups as possible inputs.

I tested the algorithms on workspace groups (with reasonable and not-so-reasonable data/workspaces in them), and I didn't manage to get any crash or bad failure.

:shipit:

File: Framework/Algorithms/test/CalMuonDetectorPhasesTest.h (hunk 1 above)
Comment (tom-perkins) [inline Framework/Algorithms/test/CalMuonDetectorPhasesTest.h]: I tried this and got "enter a name for the input in/out workspace" - it only seems to work if the workspaces are added to the ADS.

**Comment by tom-perkins** [comment]
Added a short note to the [wiki page](http://www.mantidproject.org/Writing_an_Algorithm) as described above

**Comment by tom-perkins** [comment]
Merged in master due to merge conflict

**Pull request merged**